src/pkcs15init/pkcs15-cardos5.c

```diff
@@ -0,0 +1,1876 @@
+		rsa->privkey.d.data = dummy_data;
+		rsa->privkey.d.len = rsa->bytes;
+	}
+}
+
+static void
+fill_in_dummy_ec_key(ec_keypair_t *ec, uint8_t *dummyQ, uint8_t *dummyD,
+    size_t kbits)
+{
+	memset(ec, 0x00, sizeof(*ec));
+
+	ec->bytes = kbits / 8;
+
+	dummyQ[0] = 0x04; /* uncompressed format */
+	ec->pubkey.ecpointQ.value = dummyQ;
+	ec->pubkey.ecpointQ.len = ec->bytes * 2 + 1;
+
+	ec->privkey.privateD.data = dummyD;
+	ec->privkey.privateD.len = ec->bytes;
+}
+
+static int
+push_obj(sc_card_t *card, const uint8_t *obj, size_t len, uint8_t *key_sha256)
+{
+	struct sc_cardctl_cardos_acc_obj_info	args;
+	uint8_t					payload_buf[128];
+	size_t					done;
+	size_t					n;
+	int					r;
+
+	if (len > UINT16_MAX) {
+		sc_log(card->ctx, "invalid len %zu", len);
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+
+	for (done = 0; done < len; done += n) {
+		buf_t payload = { payload_buf, sizeof(payload_buf), 0 };
+
+		if (done == 0) {
+			uint8_t len_hi = (uint8_t)(len >> 8);
+			uint8_t len_lo = (uint8_t)(len & 0xFF);
+			if (asn1_put_tag2(0x80, len_hi, len_lo, &payload)) {
+				sc_log(card->ctx, "asn1 error");
+				return SC_ERROR_BUFFER_TOO_SMALL;
+			}
+			args.append = 0; /* allocate new object */
+		} else
+			args.append = 1; /* appending to existing object */
+
+		n = len - done;
+		if (n > 64)
+			n = 64;
+
+		if (asn1_put_tag(CARDOS5_ACCUMULATE_OBJECT_DATA_TAG,
+		    obj + done, n, &payload)) {
+			sc_log(card->ctx, "asn1 error");
+			return SC_ERROR_BUFFER_TOO_SMALL;
+		}
+
+		args.data = payload_buf;
+		args.len = payload.bytes_used;
+
+		r = sc_card_ctl(card, SC_CARDCTL_CARDOS_ACCUMULATE_OBJECT_DATA,
+		    &args);
+		if (r != SC_SUCCESS)
+			return r;
+	}
+
+	if (key_sha256 != NULL)
+		memcpy(key_sha256, args.hash, 32);
+
+	return SC_SUCCESS;
+}
+
+/* add a key Control Reference template (CRT) to a buf_t. */
+static int
+add_key_crt(int key_id, int algo, int active, int pubkey, buf_t *b)
+{
+	uint8_t	changeable;
+	uint8_t	lifecycle;
+	uint8_t	kuq;
+	uint8_t	algo_type1;
+	uint8_t	algo_type2;
+
+	if (active) {
+		changeable = OBJPARAM_UNCHANGEABLE;
+		lifecycle = LIFECYCLE_OPERATIONAL;
+	} else {
+		changeable = OBJPARAM_CHANGEABLE;
+		lifecycle = LIFECYCLE_CREATION;
+	}
+
+	if (pubkey)
+		kuq = KUQ_ENCRYPT;
+	else
+		kuq = KUQ_DECRYPT;
+
+	if (algo == SC_ALGORITHM_RSA) {
+		algo_type1 = ALGO_TYPE_RSA;
+		algo_type2 = ALGO_TYPE_PARAM;
+	} else {
+		algo_type1 = ALGO_TYPE_EC;
+		algo_type2 = 0;
+	}
+
+	if (key_id < 0 || key_id > UINT8_MAX)
+		return -1;
+
+	if (asn1_put_tag1(CRT_DO_KEYREF, (uint8_t)key_id, b) ||
+	    asn1_put_tag3(CRT_TAG_OBJPARAM, OBJPARAM_NOBACKTRACK, changeable, 0,
+	    b) || asn1_put_tag1(CRT_TAG_CU, CU_CIPHER | CU_SIGN, b) ||
+	    asn1_put_tag1(CRT_TAG_KUQ, kuq, b) ||
+	    asn1_put_tag2(CRT_TAG_ALGO_TYPE, algo_type1, algo_type2, b) ||
+	    asn1_put_tag1(CRT_TAG_LIFECYCLE, lifecycle, b))
+		return -1;
+
+	return 0;
+}
+
+static int
+bertlv_put_tag(uint8_t tag, const uint8_t *data, size_t length, buf_t *buf)
```

Comment:
Indeed. This will hopefully be tackled along with the remaining
duplication that you pointed out.